### DISCUSSION TITLE
FSharp support like Ionide-VsCode

### DISCUSSION THREAD

**Discussion by WillEhrendreich** [nvim-neo-tree/neo-tree.nvim#882] [Ideas]
Hey all, 

I was wondering if there's a good way to implement some of the functionality of other IDE file explorers implementations for FSharp. 

FSharp is order-dependent, and so in the .fsproj file, the exact order of what gets evaluated by the compiler is set. if you want the code you're currently working on to know about something, it literally has to be higher in that file, or the file you want it to reference has to come before it in the fsproj.

I would like to be able to interact with the fsproj file from the tree, just like it works in https://github.com/ionide/ionide-vscode-fsharp

it uses https://github.com/ionide/proj-info to do this somehow. 

the main features i care about are the ordering of the fs files correctly, adding files above and below, and removing.. at least for now. I think this all might be fairly easy to implement, but I want to throw this out for people to comment on. 

what do you think?

**Comment by miversen33** [comment]
In _theory_ you can do this with the `sort_function` override (take a peek at the [readme](https://github.com/nvim-neo-tree/neo-tree.nvim#longer-example-for-packer), specifically for `sort_function`).

That said, I don't know how to selectively apply this to the filesystem source (which is what you want). Dynamically setting the sort function is probably what you need though. This smells like an enhancement request (though it may get shot down, I don't know).

**Reply by WillEhrendreich** [reply]
well i was thinking it would be maybe a separate source entirely and would somehow just be informed of whether or not you were in an fsharp project, and would set the default source to itself when you're in that and back to the regular filesystem source when you weren't. then all the other functionality can be implemented and won't have to poison the well in the general purpose file system source with all the extra niche complexity. 

yeah i concur that the sort_function is the way to do the ordering bit, but I'm not sure what the process of reading the fsproj is going to be like, much less writing to it.

**Reply by miversen33** [reply]
The great thing about Neo-tree is you can make your own external source and plug it into Neo-tree! [Here is an example of how to do so](https://github.com/nvim-neo-tree/example-source).

I highly doubt a completely different source for this would be added to Neo-tree core, it just seems like alot of work for a very small payoff (in the grand scheme of things). That's not to say that the idea may not be important to you and others in the F# community. Just that compared to the rest of the work that needs to be done in Neo-tree, and the work that would go into maintaining that, it seems like something that the F# community should consider building and integrating into Neo-tree.

Now to the idea that this source takes over as the `filesystem` source...

I don't know the best way to do this, but I wonder if what should instead happen is that the F# source would "be" the filesystem source. Upon opening Neotree, you would simply always have the F# source as the "default" (or first) source. It would check to see if there is an F# project in the selected directory. If so, it would remain source. If not, it would pass all calls through to the `filesystem` source.

Note, this is not an easy ask, but probably the most "seamless" in regards to what you are asking. You could also just have a completely F# source that you manually open when you know you are in an F# project, which would probably be the easiest.

In any case, this very much sounds like something that the F# community should take on if they are interested in this kind of functionality. Once it is complete, we could add a link to it in the extensions wiki. And of course questions during development and maintenance are more than welcome here :) We (or at least me lol) would have no issue providing guidance throughout the process.

**Reply by WillEhrendreich** [reply]
where is this extensions wiki? I couldn't find it referenced on the main page. I'd love to browse more of what other people have made, I didn't even know it existed. I think that should be more clear, right on the front readme like.. "Looking for community made Neo-Tree sources? CLICK HERE for our fancy list" lol.

**Reply by miversen33** [reply]
https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/External-Sources

There aren't a ton yet. It is right on the wiki sidebar though
